src/fu-engine-repair.c

```diff
@@ -0,0 +1,441 @@
+{
+	FuSecurityAttrs *attrs;
+	FwupdSecurityAttr *attr;
+	guint flags;
+
+	attrs = fu_engine_get_host_security_attrs(engine);
+	if (!attrs)
+		return FALSE;
+
+	attr = fu_security_attrs_get_by_appstream_id(attrs, FWUPD_SECURITY_ATTR_ID_UEFI_SECUREBOOT);
+	if (!attr)
+		return FALSE;
+
+	flags = fwupd_security_attr_get_flags(attr);
+
+	if (!g_strcmp0(action, "undo")) {
+		if (flags == FWUPD_SECURITY_ATTR_FLAG_SUCCESS) {
+			g_set_error_literal(
+			    error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_NOTHING_TO_DO,
+			    "Kernel lockdown can't be disabled when secure boot is enabled.");
+			return FALSE;
+		} else {
+			return grubby_set_lockdown(FALSE, error);
+		}
+	}
+
+	return grubby_set_lockdown(TRUE, error);
+}
+
+static gboolean
+fu_engine_repair_iommu(const gchar *action, GError **error)
+{
+	if (!is_grubby_installed(error)) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR_NOT_SUPPORTED,
+				    FWUPD_ERROR_NOTHING_TO_DO,
+				    "Grubby was not installed.");
+		return FALSE;
+	}
+
+	if (!g_strcmp0(action, "undo"))
+		return grubby_set_iommu(FALSE, error);
+
+	if (is_kernel_parameter("iommu=", error) || is_kernel_parameter("intel_iommu=", error) ||
+	    is_kernel_parameter("amd_iommu=", error)) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_NOTHING_TO_DO,
+				    "IOMMU had been already set.");
+		return FALSE;
+	}
+
+	return grubby_set_iommu(TRUE, error);
+}
+
+static gboolean
+fu_engine_repair_or_unsupport(FuEngine *engine,
+			      const gchar *appstream_id,
+			      const gchar *action,
+			      GError **error)
+{
+	FuSecurityAttrs *attrs;
+	FwupdSecurityAttr *attr;
+	GHashTable *settings = g_hash_table_new_full(g_str_hash, g_str_equal, g_free, g_free);
+
+	attrs = fu_engine_get_host_security_attrs(engine);
+	if (!attrs)
+		return FALSE;
+
+	attr = fu_security_attrs_get_by_appstream_id(attrs, appstream_id);
+	if (!attr)
+		return FALSE;
+
+	/* test fixable BIOS settings */
+	if (fwupd_security_attr_get_bios_setting_id(attr) != NULL &&
+	    fwupd_security_attr_get_bios_setting_current_value(attr) != NULL &&
+	    fwupd_security_attr_get_bios_setting_target_value(attr) != NULL) {
+		/* TODO: should we revert the bios settings? */
+		if (!g_strcmp0(action, "undo")) {
+			g_set_error_literal(error,
+					    FWUPD_ERROR_NOT_SUPPORTED,
+					    FWUPD_ERROR_NOTHING_TO_DO,
+					    "BIOS settings can't be reverted.");
+		}
+		g_hash_table_insert(
+		    settings,
+		    g_strdup(fwupd_security_attr_get_bios_setting_id(attr)),
+		    g_strdup(fwupd_security_attr_get_bios_setting_target_value(attr)));
+		if (!fu_engine_modify_bios_settings(engine, settings, FALSE, error))
+			return FALSE;
+
+		return TRUE;
+	}
+
+	g_set_error_literal(error,
+			    FWUPD_ERROR_NOT_SUPPORTED,
+			    FWUPD_ERROR_NOTHING_TO_DO,
+			    "Repair item is not supported.");
+	return FALSE;
+}
+
+gboolean
+fu_engine_repair_do_undo(FuEngine *self, const gchar *key, const gchar *value, GError **error)
```

Comment:
This whole function I think is begging the strings to be converted into an integer/enum equivalent and then match using switch/case or a simple if/else for supported vs unknown.

Reply:
Okay.
I will
1. Convert the parameters (actions) to an enum and then
2. perform the do/undo using a switch case.

Reply:
Should we convert the AppStream ID to the enum? I could get the Appstream ID from the security attribute and lookup the AppStream ID from the user using a for loop. For the unsupported and BIOS fixes, they can be processed in a for loop, and for those kernel fixes and exceptions, they can be filtered before the for loop then call the corresponding function to fix them.

Reply:
Sure let's see how it looks